### PR TITLE
Prepare 3.x wp.org release

### PR DESCRIPTION
**Pull request by dominic-ks**
OK, I think the time to merge the current version of this plugin that is live on WordPress.org and the latest master version on GitHub has come, with a view that subsequently the merged version will go live on WordPress.org, and from that point onwards, they shall remain in sync.

This will look like there's a lot of changes here, but really there are only a few things to consider, see summary below of the high level changes with files that contain changes to support each.

Merging these changes with the master branch doesn't change any of the behaviour of the latest version, but does bring the two in line in readiness to go to out.

**Automation to publish new versions to wp.org with GitHub actions**

NB. this is already in use and the last couple of versions went out using actions.

 - .distignore
 - .github/workflows/asset-readme-update.yml (allows updates to assets and readme without new versions)
 - .github/workflows/deploy.yml (deploys new tags as new versions)
 - *.png (all the existing wp.org assets)
 
 **The changes made in order to reinstate the plugin to wp.org after the security warning relating to firebase**

@sun I recall you had some comments previously about the use of `wp_kses`on static strings, however I cannot find those now so have made now changes to that approach...

 - class-auth.php
 - class-devices.php
 
 **Changes to provide alerts and warnings to existing users about breaking changes in V3**
 - class-plugin-update.php
 - class-setup.php
 - composer.json (added a lib to manage persisting the dismissal of the warnings)
 - jwt-auth.php
 - readme.txt

### CODE REVIEW

**Comment by sun** [comment]
Awesome. I'll try to review this tomorrow or early next week. 🙏

**Comment by dominic-ks** [comment]
> Has this been abandoned?

I hope not! Just waiting for people to be available to review it, hopefully @sun will have some time soon!

File: readme.txt
Comment (sun) [inline readme.txt:45]: If we're shooting for a wp.org release now, then we should phrase all of this in the past tense. ❤️ 

```suggestion
When updating from v2 to v3, familiarise yourself with its changes to ensure that your site continues to work as expected:

= New: Refresh tokens ([docs](https://github.com/usefulteam/jwt-auth#refreshing-the-access-token)) =

Key changes:

- Default JWT access token expiry time has been reduced from 7 days to 10 minutes.
- On expiry of a JWT, clients need to retrieve a new access token using the [refresh token as described here](https://github.com/usefulteam/jwt-auth#refreshing-the-access-token).
- To retain the 7 day expiry time, use the hook `jwt_auth_expire`.

= Removed Whitelist =

Key changes:

- You no longer need to whitelist REST paths from other plugins with the hook `jwt_auth_whitelist`. You can remove the hook.
- Instead, custom REST API routes should have access requirements specified with the [permissions callback](https://developer.wordpress.org/rest-api/extending-the-rest-api/adding-custom-endpoints/#permissions-callback) when it is registered.
- This means that if a route requires authentication, any authentication method can be used and this should reduce conflicts between this and other plugins. See [this discussion](https://github.com/usefulteam/jwt-auth/pull/60) for further information.
```

File: readme.txt
Comment (sun) [inline readme.txt:793]:
```suggestion
- Added automated asset updates from GitHub.

= 2.1.5 =
- Removed dev and build files from distribution.

= 2.1.4 =
- Added update warning and information relevant to updating to version 3.
```

**Review by sun — APPROVED** [review]
I'm terribly sorry for taking so long to review this - I hope things will get better for me soon. 🙏 

This looks fine aside from the bootstrap code loading issue.  But that and everything else mentioned in my review is rather minor.  I'm already (pre-)approving even though I think we should apply the suggestions and do the mentioned cleanups.

Note that I also just created https://github.com/usefulteam/jwt-auth/issues/117, which keeps on popping up in our site's profiling in production.  It's only a PHP warning, so we could release 3.0 without a fix for it, but of course, several people might run into it, too.

File: class-plugin-update.php
Comment (dominic-ks) [reply class-plugin-update.php]: Up to you I guess, I'd just seen others had done it and I'm always after fame so thought I would too!

File: jwt-auth.php
Comment (dominic-ks) [reply jwt-auth.php]: I have moved this to the if statement you mentioned for now, not sure if that's correct re: `is_admin()` returning true for REST API, maybe you're thinking of `wp_ajax`?

File: class-plugin-update.php
Comment (dominic-ks) [reply class-plugin-update.php]: Have updated the names.

File: class-plugin-update.php
Comment (dominic-ks) [reply class-plugin-update.php]: Thanks, have updated those, not sure where they came from.

**Comment by dominic-ks** [comment]
@sun Thanks for taking the time to review this so thoroughly, I've accepted your suggestions and addressed the other points as well. I will also take a look through those other PRs you mentioned before merging those _and_ this one, then hopefully we can move on!

**Comment by sun** [comment]
All mentioned PRs have been merged. 👌 

Question: When are we actually updating the changelog?  Only when tagging/creating a release?  Or should this be done with every PR?

**Comment by dominic-ks** [comment]
It's a good question, I guess we don't have a policy, nor one that says when we will tag a new release. On updating the changelog specifically, I think it makes sense to update it on every PR merge, not only will it be easier since we won't have to look back, but also it means that the change log on the master branch will actually reflect what has been added. Perhaps under the title of "current master" or similar, which can then just be changed to the version number once a release is tagged?

Otherwise, I think this is ready to be merged, and I think we should then plan for this to finally go to wp.org. Once merged, the next tag will deploy automatically.

**Comment by sun** [comment]
Yeah. Typically people are using the major version as the heading; i.e., "3.x.x".  After merging changes, their compatibility decides what the next version number is going to be.

This would mean that we'd need to do this before tagging a release.

**Comment by sun** [comment]
In essence, we should add the following to the changelog in the readme.txt (outside of diff context, so I sadly cannot suggest):

```
= 3.0.2 =
- Fix: Do not revalidate authentication headers if a valid user was determined already. (#75)
- Fix: Added debugging timeframe before purging refresh tokens. (#93)
- Fix: Fixed unnecessary user account lookup for device listing on user profile page. (#84)
- Fix: Added more granular refresh token validation error messages. (#78)
- Fix: Added integration for new CORS filter hook rest_allowed_cors_headers in WordPress 5.5.0. (#97)
- Fix: Updated Guzzle to v7.8.1 (used in tests only). (#112)

```
I'd recommend to continue the new versions and actually release the version 3.0.2 to wp.org, so that everyone who was using the plugin from GitHub will see a difference to their currently installed version.

**Pull request merged**
